### PR TITLE
Project Mjölnir: Part 2 - Controller Applet

### LINKED ISSUE #2606: Synchronization error in applet state changed event softlock
This was caused by #2416 however as #2416 corrects behavior to HW-accurate, this appears to be an issue in how the AppletStateChanged event is fired.

This event is used to signal completion of an applet, for example software keyboard or controller.

This is the cause of many games crashing or softlocking when attempting to use swkbd currently. My hypothesis is that the minor sync changes in the above PR cause the game to either wait on an event that will not signal or to miss the signalling of an event (or perhaps receive the signal but not stop waiting).

Will do further investigation, as this also causes an infinite loop/softlock with controller applet (#1940)

<meta name="titleids"  content="010003F003A34000,0100187003A36000,01007EF00011E000">

### LINKED ISSUE #3465: Megaman Zero/ZX Legacy Collection, no input method works
[yuzu_log.zip](https://github.com/yuzu-emu/yuzu/files/4258147/yuzu_log.zip)

Game won't get past the splash screen due to input problems.
Tried most possible combinations.

### LINKED ISSUE #3674: Multiplayer not working in super mario maker 2
**Mario Maker 2:**
Pressing on the "Play together" option does nothing but on real HW it says press L+R on your controller.
yuzu:
Screenshot (1153)
Real HW:
Screenshot (1156)
yuzu log:
[Mario_Maker_2.zip](https://github.com/yuzu-emu/yuzu/files/4483515/Mario_Maker_2.zip)

### LINKED ISSUE #3989: Hyrule Warriors: Definitive Edition - Controller 2 not recognized
Hey guys, I was attempting to play Hyrule Warriors: Definitive Edition over Parsec with a friend, but it seems that the game doesn't recognize a second controller ever being plugged-in.

Normally, you need to press the Left Stick in the character selection menu to enable Coop, but for some reason, even when 2 controllers are configured properly, the game won't detect inputs from Controller 2.

The bug is present on the latest yuzu Mainline (265) :
dsa

### PR DESCRIPTION
**Pull request by Morph1984**
# Project Mjölnir: Part 2 - Controller Applet

This is Part 2 of a multipart rewrite to yuzu's HID (Human Interface Device) frontend and backend implementations.

This part implements a HLE controller selector applet, also known as applet 0x0C.

The controller applet is used extensively in many games to set up additional controllers for multiplayer.
Without it, games are not able to properly detect multiple controllers despite having them connected as it collects the result from the applet itself.

I and @Its-Rei teamed up once more to create a controller applet that closely resembles the Nintendo Switch's native UI while providing access to the input configuration if needed.

## Qt Frontend

The Qt Frontend Applet is designed to resemble the Nintendo Switch's native UI implementing all of its features such as:

- Controller Support Icons
- Controller Identification (Border) Color
- Controller Explain Text

The applet UI automatically adapts to any parameters passed in from the backend, showing anywhere from 1 to 8 players depending on the maximum amount of allowed players.
Access to the input configuration is available and also shows the maximum amount of allowed players.

Here are comparisons of our applet UI compared to the Nintendo Switch:

### Super Mario Odyssey, 2 Player Mode with Explain Text
Switch:
supermarioodyssey-switch

Yuzu:
supermarioodyssey-yuzu

### Mario Kart 8 Deluxe, 4 Player Mode with Identification (Border) Color
- Note that this game still softlocks after the applet, the softlock is unrelated.

Switch:
mariokart8deluxe-switch

Yuzu:
mariokart8deluxe-yuzu

### Kirby Star Allies, 4 Player Mode with Identification (Border) Color and variable player amounts (1-4)
Switch:
kirbystarallies-switch

Yuzu:
kirbystarallies-yuzu

### Super Mario Maker 2, Play Together with Identification (Border) Color, Explain Text and variable player amounts (2-4)
Switch:
supermariomaker2-switch

Yuzu:
supermariomaker2-yuzu

### Shipped, Allows up to 8 Players
Switch:
shipped-switch

Yuzu:
shipped-yuzu

### Hyrule Warriors, Controller Applet on boot allowing only a single player.
Switch:
hyrulewarriorsde-switch

Yuzu:
hyrulewarriorsde-yuzu

### Captain Toad Treasure Tracker, 2 Player Assist Mode with Left and Right Joycons
- Note that only a pair of left and right joycons work here, using only left or only right joycons will **NOT** work.

Switch:
captaintoadtreasuretracker-switch

Yuzu:
captaintoadtreasuretracker-yuzu

### Known issues
- The Applet UI may seem to lag/slowdown in certain games, its cause is uncertain at the moment.

Like [Part 1](https://github.com/yuzu-emu/yuzu/pull/4530), the following UI elements are placeholders for any present or future PR:

- Vibration percentage spinbox
- Motion configuration button

## SDL Frontend (yuzu-cmd)

The SDL frontend will have a fallback "auto" applet which deduces the best configuration based on the parameters passed in from the backend.

The following heuristic is used:

1. Disconnect all currently connected controllers (including handheld).
2. Connect only the minimum number of players that are required with a minimum of 1 connected player since some games allow a minimum of 0.
3. Connect controllers based on the following priority list from highest to lowest priority (within the given parameters, of course):
- Pro Controller -> Dual Joycons -> Left Joycon/Right Joycon -> Handheld
4. Assign left joycons to even player indices and right joycons to odd player indices.
- We do this since Captain Toad Treasure Tracker expects a left joycon for Player 1 and a right Joycon for Player 2 in 2 Player Assist mode.

This closes #2606, closes #3465, closes #3674, and closes #3989
This supersedes #1940

### CODE REVIEW

File: src/core/frontend/applets/controller.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                npad.AddNewControllerAt(
+                    npad.MapSettingsTypeToNPad(Settings::ControllerType::RightJoycon), index);
+            }
+        } else if (index == 0 && parameters.enable_single_mode && parameters.allow_handheld &&
+                   !Settings::values.use_docked_mode) {
+            // We should *never* reach here under any normal circumstances.
+            npad.AddNewControllerAt(npad.MapSettingsTypeToNPad(Settings::ControllerType::Handheld),
+                                    index);
+        } else {
+            UNREACHABLE_MSG("Unable to add a new controller based on the given parameters!");
```
Comment (lioncash) [inline src/core/frontend/applets/controller.cpp:71]: `"common/assert.h"` needs to be included.

File: src/core/hle/service/am/applets/controller.cpp [hunk 2]
```diff
@@ -0,0 +1,206 @@
+// Copyright 2020 yuzu Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include <algorithm>
```
Comment (lioncash) [inline src/core/hle/service/am/applets/controller.cpp:5]: `<cstring>` needs to be included.

File: src/core/hle/service/am/applets/controller.cpp [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+#include "core/core.h"
+#include "core/frontend/applets/controller.h"
+#include "core/hle/service/am/am.h"
+#include "core/hle/service/am/applets/controller.h"
+#include "core/hle/service/hid/controllers/npad.h"
+
+namespace Service::AM::Applets {
+
+// This error code (0x183ACA) is thrown when the applet fails to initialize.
+[[maybe_unused]] constexpr ResultCode ERR_CONTROLLER_APPLET_3101{ErrorModule::HID, 3101};
```
Comment (lioncash) [inline src/core/hle/service/am/applets/controller.cpp:17]: `core/hle/result.h` needs to be included.

File: src/core/hle/service/am/applets/controller.cpp [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    library_applet_version = LibraryAppletVersion{common_args.library_version};
+
+    const auto private_arg_storage = broker.PopNormalDataToApplet();
+    ASSERT(private_arg_storage != nullptr);
+
+    const auto& private_arg = private_arg_storage->GetData();
+    ASSERT(private_arg.size() == sizeof(ControllerSupportArgPrivate));
+
+    std::memcpy(&controller_private_arg, private_arg.data(), sizeof(ControllerSupportArgPrivate));
+    ASSERT_MSG(controller_private_arg.arg_private_size == sizeof(ControllerSupportArgPrivate),
```
Comment (lioncash) [inline src/core/hle/service/am/applets/controller.cpp:70]: `common/assert.h` needs to be included.

File: src/core/hle/service/am/applets/controller.cpp [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+                        controller_user_arg_new.identification_colors.begin(),
+                        controller_user_arg_new.identification_colors.end()),
+                    std::vector<ExplainText>(controller_user_arg_new.explain_text.begin(),
+                                             controller_user_arg_new.explain_text.end()));
+            }
+        }();
+
+        is_single_mode = parameters.enable_single_mode;
+
+        LOG_DEBUG(Service_HID,
```
Comment (lioncash) [inline src/core/hle/service/am/applets/controller.cpp:152]: `common/logging/log.h` needs to be included.

File: src/core/hle/service/am/applets/controller.h [hunk 6]
```diff
@@ -0,0 +1,120 @@
+// Copyright 2020 yuzu Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <array>
```
Comment (lioncash) [inline src/core/hle/service/am/applets/controller.h:7]: `<vector>` needs to be included.

File: src/core/hle/service/am/applets/controller.h [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+#include "core/hle/result.h"
```
Comment (lioncash) [inline src/core/hle/service/am/applets/controller.h:9]: `common/common_funcs.h` and `common/common_types.h` need to be included.

File: src/yuzu/applets/controller.cpp [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+#include "core/hle/lock.h"
+#include "core/hle/service/hid/controllers/npad.h"
+#include "core/hle/service/hid/hid.h"
+#include "core/hle/service/sm/sm.h"
+#include "ui_controller.h"
+#include "yuzu/applets/controller.h"
+#include "yuzu/configuration/configure_input_dialog.h"
+#include "yuzu/main.h"
+
+constexpr std::array<std::array<bool, 4>, 8> led_patterns = {{
```
Comment (lioncash) [inline src/yuzu/applets/controller.cpp:18]: This should be within the anonymous namespace below.

File: src/yuzu/applets/controller.cpp [hunk 9]
```diff
@@ -0,0 +1,600 @@
+// Copyright 2020 yuzu Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include <algorithm>
+
+#include "common/string_util.h"
```
Comment (lioncash) [inline src/yuzu/applets/controller.cpp:7]: `common/assert.h` needs to be included.

**Review by bunnei — APPROVED** [review]

**Pull request merged**